packages/devtools/lib/src/memory/memory.dart

```diff
@@ -362,38 +650,291 @@ class MemoryScreen extends Screen with SetStateMixin {
+        row.className,
+        hashCodeResult.valueAsString,
+      ));
+
+      if (memoryPath.isLastInBound) {
+        // Re-construct memory navigation and display.
+        history.clear();
+        memoryPath.addPathsLinks(history, _handleHistoryClicks);
+
+        history..clazz('history-navigation', removeOthers: true);
+      }
+    }
+  }
+
+  void _handleHistoryClicks(CoreElement element) {
+    // Handle clicking in the history links.
+    if (element.tag == 'SPAN' && element.hasClass('history-link')) {
+      final attrs = element.attributes;
+
+      final int dataIndex = int.parse(attrs[NavigationState.dataIndex]);
+
+      final String dataClass = attrs[NavigationState.dataClass];
+      final String dataField = attrs.containsKey(NavigationState.dataField)
+          ? attrs[NavigationState.dataField]
+          : '';
+      final String dataHashCode =
+          attrs.containsKey(NavigationState.dataHashCode)
+              ? attrs[NavigationState.dataHashCode]
+              : '';
+
+      final NavigationState state = memoryPath.get(dataIndex);
+
+      // The clicked link's attributes and real NavigationState should match.
+      assert(dataClass == state.className &&
+          dataField == state.field &&
+          dataHashCode == state.objectHashCode);
+
+      // Prune remove this state and to the end as well.
+      memoryPath.remove(state);
+
+      if (state.isClass) {
+        _selectClass(state.className);
+      } else if (state.isInstance) {
+        selectClassInstance(state.className, state.objectHashCode);
+      } else if (state.isInbound) {
+        // Same as selecting the class instance but record the field, don't
+        // need to match following the refs.
+        memoryPath.fieldReference = state.field;
+        selectClassInstance(state.className, state.objectHashCode);
+      } else {
+        assert(false, 'Unknown NavigationState');
+      }
+
+      history.clear();
+      Timer(const Duration(milliseconds: 100), () {
+        if (!memoryPath.isLastInBound) {
+          history..clazz('history-navigation-hidden', removeOthers: true);
+        } else {
+          memoryPath.addPathsLinks(history, _handleHistoryClicks);
+        }
+      });
+    }
+  }
+
+  void hoverInstanceAllocations(HoverCellData<InstanceSummary> hover) async {
+    if (hover.cell == null) {
+      // Hover out.
+      instancePopup.clear(); // Remove all children.
+      instancePopup.display = 'none';
+      return;
+    }
+
+    // Entering Hover again?
+    if (instancePopup.element.children.isNotEmpty) return;
+
+    final CoreElement ulElem = ul();
+    final InboundReferences refs =
+        await getInboundReferences(hover.data.objectRef, 1000);
+    ulElem.add(li(c: 'allocation-li-title')
+      ..add([
+        span(text: 'Allocated', c: 'allocated-by-class-title'),
+        span(text: 'Referenced', c: 'ref-by-title')
+      ]));
+
+    for (InboundReference element in refs.elements) {
```

Comment:
This is pretty complex business logic. Can we move it to a file that be tested without pulling in dart:html dependencies?
The way we have a separate controller that can be tested with dart:vm unit tests has worked nicely for the profiler and inspector code. it will also help with the future move to flutter_web

Reply:
Yep, moved the computations to MemoryController.  Need to have a MemoryModel to separate out all the data in MemoryScreen.